Add macho header parser

jorgerodrigoleclercq opened the pull request:
- Create directories `Macho` inside `include/parser` and `lib/parser`
- Create files `macho_header.h` and `macho_header.cpp` inside `Macho` directories
- Create `CMakeLists.txt` inside `lib/parser/Macho`
- Implement `MachoHeader::parse_header`, which reads and prints a macho header
- Modify `shuriken/CMakeLists.txt` to include MACHO functionalities

jeppojeps committed:
Update README.md

added step-by-step

badumbatish committed:
Simple opt pass

Introduce opt, not working until smali branch

Abstracts away opt

jorgerodrigoleclercq committed:
Add macho header parser

- Create directories `Macho` inside `include/parser` and `lib/parser`
- Create files `macho_header.h` and `macho_header.cpp` inside `Macho` directories
- Create `CMakeLists.txt` inside `lib/parser/Macho`
- Implement `MachoHeader::parse_header`, which reads and prints a macho header
- Modify `shuriken/CMakeLists.txt` to include MACHO functionalities

badumbatish commented on shuriken/include/shuriken/parser/Macho/macho_header.h:
```diff
@@ -0,0 +1,53 @@
+namespace shuriken::parser::macho {
+    class MachoHeader {
+    public:
+#pragma pack(1)
```
can i ask why we're using pack here?

Fare9 replied:
This is correct to avoid some issue in case some padding is included in the structure. When reading from a data stream and you read a full structure, you can have a byte mismatch if you have some padding.

badumbatish commented on shuriken/CMakeLists.txt:
```diff
@@ -133,6 +134,35 @@ if (DEX)
 endif ()
 
+# Compile the object files for MACHO
+if (MACHO)
```
maybe we should add this option to the ci, test if it's breaking anything

Fare9 replied:
yep, we need to add -DMACHO=ON to avoid Shuriken-dump to break, it shouldn't be difficult to add it in this PR.

Fare9 replied:
let's leave it for the moment like this. AFter the release we will add it to the CI.

badumbatish committed:
Fix unused var in mjolnir op graph gen

Fare9 commented on shuriken/lib/parser/Macho/macho_header.cpp:
```diff
@@ -0,0 +1,32 @@
+    // read the macho header
+    stream.read_data<machoheader_t>(machoheader, sizeof(machoheader_t));
+
+    // print the macho header
```
a `parse_*` function should not contain the printing of the fields, this corresponds to another function where the information is dumped, but if you add it here, the data will be printed every time you parse any mach-o header.

JorgeRodrigoLeclercq replied:
That's right, I'll divide it into two functions and add some checkings. Thanks!

Fare9 commented on shuriken/lib/parser/Macho/macho_header.cpp:
```diff
@@ -0,0 +1,32 @@
+    std::cout << "Flags: 0x" << std::hex << machoheader.flags << std::endl;
+    std::cout << "Reserved: 0x" << std::hex << machoheader.reserved << std::endl;
+
+    log(LEVEL::INFO, "Finished parsing header");
```
If there's any check to do in the structure, you can do it right before this call to `log`.

JorgeRodrigoLeclercq commented on shuriken/CMakeLists.txt:
```diff
@@ -133,6 +134,35 @@ if (DEX)
+    )
+
+    # Macho files
+    add_subdirectory(../../Shuriken-Analyzer/shuriken/lib/parser/Macho)
```
You are right, I messed up. I'll change it to `add_subdirectory(lib/parser/Macho)` just as it is for the DEX parser

JorgeRodrigoLeclercq commented on shuriken/CMakeLists.txt:
```diff
@@ -133,6 +134,35 @@ if (DEX)
+    message(STATUS "Generating Shuriken MACHO object files")
+
+    add_library(macho-obj
+            OBJECT
```
To be honest I just copied that from the `if (DEX)` block, I'll take a look into it

Fare9 committed:
Fixed issue creating basic blocks

badumbatish committed:
Fix Include llvm as system,

- Fix Include llvm as system, this allow for more warning enabled

- Remove unused function in llvm

badumbatish committed:
Scaffoling the mjolnir to smali pass (#109)

* scaffolding dalvik

Correctly provide include directory, make new rewriter for lowering

Working on rewritter

Opted for cpp based rewriting

Implemented fallthrough lowering(not sure if its working), add basic pass

Pass manager wrapper for lowering

Lowering pass partially runnable, bug in mjolnir

Lowering partially, remove nop first

Include llvm in system

Remove bogus libs from Jasmine

Add logical result

Remove dalvik dialect (useless), finalize rebase from main

Rename from lower to opt

let opt go into passes, mjolnir to smali into transform

Resolve namespace conflict by swapping include order

Move most of opt to .cpp for faster compilation time

Modify cmdline opt for better control of different phases

Minimum barebone support for lowering achieved on hello world

Lowering half way through

Change std::string to std::stringview

Rebase on current main, each class is a module

Refactor getting v{} and p{}, introduced access_flags_to_string

WIP in lowering

* Modified code to add the flags to the method declaration (#133)

* Added flags into MethodOp
* Modified code for including the method flags in the lifting process
* Added the flags into the lowering process

---------

Co-authored-by: Fare9 <eduardo.blazquez@uc3m.es>

badumbatish committed:
FallThroughOp is now of BranchOpInterface (#134)

jorgerodrigoleclercq committed:
Fix errors in CMakeLists.txt and ajdust Macho-O header handling
*Fix errors in CMakeLists.txt and adjust Mach-0 header handling

- Add `print_macho_header` in `shuriken-dump/shuriken-dump.cpp`, which prints a Mach-O's header in the terminal
- Modify `MachoHeader::parse_header` to only parse a Macho's header
- Create tests for Mach-O headers in `shuriken/tests/macho/

robert-yates committed:
fix type warning for windows build (#143)

Fare9 committed:
Add test on CI for Windows compilation (#144)

* Start cmake-windows.yml

* Adding flag to avoid specific warnings from msvc

* Avoided warnings for externals

* Ignore tests

* Ignore tests

* avoided errors in uninstallation check

* avoided shuriken-dump test

* Check only that the folder is empty

* Checked install for the moment only

Fare9 committed:
145 an unicodedecodeerror occurs when getting strings using the python binding (#150)

* Fixed string representation for APKs

* Changed the errors=replace to errors=backslashreplace, added one method to obtain the string as a bytearray

Fare9 committed:
Improved setup.py adding an option to install as user to avoid instal… (#151)

* Improved setup.py adding an option to install as user to avoid installing on a system folder

* updated URL for shuriken

* Added user installation to python installation

* Small fix python-install.yml

* Removed virtual env for user installation

* Upgraded setup.py version

* Shorten python ci/cd of pip install --user

* Remove check of user in .venv/

---------

Co-authored-by: badumbatish <tanghocle456@gmail.com>

Fare9 committed:
Added .def files to the installation (#154)

Fare9 committed:
Improvements c api code (#155)

* Organized code in the C API

* Work In Progress with the tests and the C API

* Cleaned code and fixed issue with handlers from disassembled methods

* Fixed compiler warning

* Fixed last compiler warnings

lukhio committed:
Access DEX header in C API (#156)

* Add DEX header struct in C api

* Prototype for get_header() for hDexContext

* Add DEX header getter

* Switching DEX header checksum type to std::uint32_t

* Store header info in opaque struct

* Fix header tests

* Missing function declaration

* DEX header access from APK C API too

Fare9 commented:
Code seems now OK to me, I would add a conditional compilation in `shuriken-dump`, you will need to define a variable in CMakeLists.txt, and then check if the variable is defined with 

```
#ifdef MACHO
#endif
```

Whatever is between those pragmas is compiled only if the variable is defined, in this way, the compilation will not crash. Also remember that you have to rebase to retrieve the last commits in CMakeLists.txt

jorgerodrigoleclercq committed:
Merge branch 'main' into macho-files
